src/gba/hardware-extensions.c

```diff
@@ -0,0 +1,339 @@
+/* Copyright (c) 2021 Jeffrey Pfau
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+#include <mgba/internal/gba/hardware-extensions.h>
+#include <mgba/internal/gba/memory.h>
+#include <mgba/internal/gba/io.h>
+#include <mgba/internal/gba/gba.h>
+
+
+enum HWEX_RETURN_CODES {
+    HWEX_RET_OK = 0,
+    HWEX_RET_WAIT = 1,
+
+    // Errors
+    HWEX_RET_ERR_UNKNOWN = 0x100,
+    HWEX_RET_ERR_BAD_ADDRESS = 0x101,
+    HWEX_RET_ERR_INVALID_PARAMETERS = 0x102,
+    HWEX_RET_ERR_WRITE_TO_ROM = 0x103,
+    HWEX_RET_ERR_ABORTED = 0x104,
+    HWEX_RET_ERR_DISABLED = 0x105,
+    HWEX_RET_ERR_DISABLED_BY_USER = 0x106
+};
+
+#define SIMPLIFY_HWEX_REG_ADDRESS(address) ((address - REG_HWEX0_CNT) >> 1)
+
+const uint16_t extensionIdByRegister[] = {
+    // More RAM
+    [SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_CNT)] = HWEX_ID_MORE_RAM,
+    [SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_RET_CODE)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P0_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P0_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P1_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P1_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P2_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P2_HI)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P3_LO)] = HWEX_ID_MORE_RAM,
+	[SIMPLIFY_HWEX_REG_ADDRESS(REG_HWEX0_P3_HI)] = HWEX_ID_MORE_RAM,
+};
+
+static uint16_t GetExtensionIdFromAddress(uint32_t address) {
+    return extensionIdByRegister[SIMPLIFY_HWEX_REG_ADDRESS(address)];
+}
+
+#undef SIMPLIFY_HWEX_REG_ADDRESS
+
+// CNT flags
+// Writable
+#define HWEX_CNT_FLAG_CALL 1
+#define HWEX_CNT_ALL_WRITABLE (HWEX_CNT_FLAG_CALL)
+// Read only
+#define HWEX_CNT_FLAG_PROCESSING (1 << 15)
+
+struct HardwareExtensionHandlers {
+    uint16_t (*onCall)(struct GBA* gba);
+    bool (*onAbort)(void);
+};
+
+static const struct HardwareExtensionHandlers extensionHandlers[];
+
+void GBAHardwareExtensionsInit(struct GBAHardwareExtensions* hwExtensions) {
+	hwExtensions->enabled = false;
+
+    hwExtensions->userEnabled = false;
+    memset(hwExtensions->userEnabledFlags, 0, sizeof(hwExtensions->userEnabledFlags));
+
+    memset(hwExtensions->memory, 0, sizeof(hwExtensions->memory));
+}
+
+static uint32_t GetHwExMemoryIndex16FromAddress(uint32_t address) {
```

Comment:
```suggestion
static uint32_t _getHwExMemoryIndex16FromAddress(uint32_t address) {
```

Why does this function need to exist anyway,? Its name is practically as long as its own code.

Reply:
I deleted it, don't know why I made it in the first place...